src/lib/helpers.ts

```diff
@@ -50,3 +50,69 @@ export async function createDirectoryIfNotExists(path: string) {
     }
   }
 }
+
+export function hasEveryOverlap<T>(array1: T[], array2: T[]): boolean {
+  if (!Array.isArray(array1) || !Array.isArray(array2)) {
+    return false;
+  }
+
+  return array1.every(element => array2.includes(element));
+}
+
+export function hasAnyOverlap<T>(array1: T[], array2: T[]): boolean {
+  if (!Array.isArray(array1) || !Array.isArray(array2)) {
+    return false;
+  }
+
+  return array1.some(element => array2.includes(element));
+}
+
+export function ignoredNamespaceConflict(ignoreNamespaces: string[], bindingNamespaces: string[]) {
+  return hasAnyOverlap(bindingNamespaces, ignoreNamespaces);
+}
+
+export function bindingAndCapabilityNSConflict(bindingNamespaces: string[], capabilityNamespaces: string[]) {
+  if (!capabilityNamespaces) {
+    return false;
+  }
+  return capabilityNamespaces.length !== 0 && !hasEveryOverlap(bindingNamespaces, capabilityNamespaces);
+}
+
+export function generateWatchNamespaceError(
+  ignoredNamespaces: string[],
+  bindingNamespaces: string[],
+  capabilityNamespaces: string[],
+) {
+  let err = "";
+
+  // check if binding uses an ignored namespace
+  if (ignoredNamespaceConflict(ignoredNamespaces, bindingNamespaces)) {
+    err += "Binding uses a Pepr ignored namespace.";
+  }
+
+  // ensure filter namespaces are part of capability namespaces
+  if (bindingAndCapabilityNSConflict(bindingNamespaces, capabilityNamespaces)) {
+    err += "Binding uses namespace not governed by capability.";
+  }
+
+  // add a space if there is a period in the middle of the string
+  return err.replace(/\.([^ ])/g, ". $1");
+}
+
+// namespaceComplianceValidator ensures that capability bindinds respect ignored and capability namespaces
+export function namespaceComplianceValidator(capability: CapabilityExport, ignoredNamespaces?: string[]) {
+  const { namespaces: capabilityNamespaces, bindings, name } = capability;
+  const bindingNamespaces = bindings.flatMap(binding => binding.filters.namespaces);
+
+  const namespaceError = generateWatchNamespaceError(
+    ignoredNamespaces ? ignoredNamespaces : [],
+    bindingNamespaces,
+    capabilityNamespaces ? capabilityNamespaces : [],
+  );
+  if (namespaceError !== "") {
+    console.error(
+      `Error in ${name} capability. A binding violates namespace rules. Please check ignoredNamespaces and capability namespaces: ${namespaceError}`,
+    );
+    process.exit(1);
```

Comment:
I'm generally not a huge fan of seeing "process.exit()" calls inside helpers / library code / anywhere _outside_ of the CLI code because:
- it doesn't allow the caller an opportunity to correct issues before blowing up the thread,
- it doesn't allow the caller to a way to "do things" (e.g. close child procs, clean up files, and so on) before the process dies, and
- it doesn't allow the caller to modify the exit codes / content or syntax of messages that are output (in the case they have a certain style of log output, etc.).

Generally, things like this are better handled by throwing an error rather than killing the process because it allows the caller the opportunity to act on those points (and probably more).  How hard would it be to push the responsibility to exit up to a higher (preferably CLI-owned) layer and just have this func throw to quit?

Reply:
I hear you, does it make a different that this is only called from the `cli` in `npx pepr build`?

![image](https://github.com/defenseunicorns/pepr/assets/1096507/a2320b14-b641-4ee4-b8dd-aa2c0f99e870)

Reply:
What are your thoughts?

Reply:
Does it make a difference that it's only called from the CLI..?  Kinda, but kinda not. :thinking: 

I think it's mostly about optionality & not _implicitly_ tying that function to the CLI  / assuming the caller is architecturally-empowered kill the whole execution environtment, more than anything.  Because that function isn't declared in a CLI-specific file / namespace or anything -- its declared in src/lib which is exactly the same place we'd put code that we expected to employ in both CLI and non-CLI contexts, right?  Admittedly, we're not making a great distinction between CLI & non-CLI in our source artifacts, but... from that location, there's not really anything even suggesting to a future caller that they might be importing a function that will explode their process without the chance of recovery.

So, is it a show stopper?  Nah.  But is it good style / a pattern I think we'd want to propagate throughout our code base?  Also probably _not_.  And any code that makes it into main has at least the potential to be copied by new / future devs, so... you know.  Kind of a judgement call.  :shrug: 

Better to just make it exit with a throw and let that propagate up to the top level & kill the process IMO.  At least that way we leave the option for someone in the future to use this call in a way that allows recovery / retry / etc.